### PR TITLE
Fix - Distance Request Optimistic Data (26518)

### PR DESCRIPTION
**Pull request by jeet-dhandha**
### Details

- Wrong optimistic data was sent.

### Fixed Issues

$ https://github.com/Expensify/App/issues/26518
PROPOSAL: https://github.com/Expensify/App/issues/26518#issuecomment-1703270319

### Tests

1. Request money for a distance request.
2. After adding all the necessary data and confirming the request.
3. Verify that Amount is displayed in Message instead of "0.00".

- [X] Verify that no errors appear in the JS console

### Offline tests

1. Request money for a distance request.
2. After adding all the necessary data and confirming the request.
3. In offline data is not shown for the users. As map service is not available in offline so all the data will update after user gets online.

### QA Steps

1. Request money for a distance request.
2. After adding all the necessary data and confirming the request.
3. Verify that Amount is displayed in Message instead of "0.00".

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/78416198/4da6f8ab-3350-48b5-9b8b-f5ae9e50a718

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/78416198/8b919752-5750-423f-8b85-aeb656d5f5a8

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/78416198/3e3ec122-0214-4940-8d8f-2d6025c6acba

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/78416198/733ae426-da48-4c07-a26f-f5b8b218d4bc

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/78416198/8f3b77d3-7ac9-41d0-b920-7550d11e4361

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/78416198/05d358b3-32c2-4925-9318-e6c179182c57

</details>

### CODE REVIEW

File: src/libs/TransactionUtils.js
Comment (hayata-suenaga) [inline src/libs/TransactionUtils.js:267]: @jeet-dhandha why is this change necessary?

Comment (jeet-dhandha) [reply src/libs/TransactionUtils.js:267]: Transaction's `reportID` was set as number and due to which the filter was not working. As passed id was a string type.

File: src/components/MoneyRequestConfirmationList.js
Comment (jeet-dhandha) [reply src/components/MoneyRequestConfirmationList.js:342]: Let me check that too 👍

File: src/libs/TransactionUtils.js
Comment (hayata-suenaga) [reply src/libs/TransactionUtils.js:267]: where do we use `getAllReportTransactions`?

**Review by hayata-suenaga — DISMISSED** [review]

File: src/libs/TransactionUtils.js
Comment (neil-marcellini) [reply src/libs/TransactionUtils.js:267]: Let's fix the root cause and make sure reportID on the transaction is set as a string. All reportIDs should be strings.

Comment (hayata-suenaga) [reply src/libs/TransactionUtils.js:267]: where do we use `getAllReportTransactions`?

**Comment by hayata-suenaga** [comment]
ops don't merge I accidentally approved 😨

**Comment by jeet-dhandha** [comment]
https://github.com/Expensify/App/blob/8fbe6aa33ba0eb5d0aa5fb56a4c8a31fda7259f2/src/components/ReportActionItem/ReportPreview.js#L115

@hayata-suenaga Here `getTransactionsWithReceipts` uses internally.

**Comment by hayata-suenaga** [comment]
> Transaction's reportID was set as number and due to which the filter was not working. As passed id was a string type.

@jeet-dhandha 

which reportID is compared to which reportID. What are data sources being compared?

**Comment by jeet-dhandha** [comment]
Report ID of IOU and the report id available in transactions.

**Comment by jeet-dhandha** [comment]
@hayata-suenaga can you suggest me a faster way to build android ?

**Comment by hayata-suenaga** [comment]
> Let's fix the root cause and make sure reportID on the transaction is set as a string. All reportIDs should be strings.

@jeet-dhandha can you actually make sure that reportID on the transaction is stored as a number?

**Comment by hayata-suenaga** [comment]
> @hayata-suenaga can you suggest me a faster way to build android ?

Let me try test build. Are you on beta for the distance request feature?

File: src/components/MoneyRequestConfirmationList.js
Comment (jeet-dhandha) [reply src/components/MoneyRequestConfirmationList.js]: Updated code for this to use the existing `IOU.setMoneyRequestMerchant`.

**Comment by hayata-suenaga** [comment]
@jeet-dhandha test build in process [here](https://github.com/Expensify/App/actions/runs/6055108026)...

by the way, thank you for sticking with us. I know it's late over there

File: src/libs/TransactionUtils.js
Comment (jeet-dhandha) [reply src/libs/TransactionUtils.js:267]: This might need to be handled in back-end. while creating the transactions as on front-end i checked the `reportID` its in string.

**Comment by jeet-dhandha** [comment]
Please check the screen shot.

Screenshot 2023-09-02 at 5 52 37 AM

**Comment by hayata-suenaga** [comment]
> Please check the screen shot.

@jeet-dhandha thank you for the screenshot. I can see that reportID is stored as a string.

Are these transactions that has reportID as a string all coming from the distance request or are they generally stored as a string?

File: src/pages/iou/steps/MoneyRequestConfirmPage.js
Comment (jeet-dhandha) [reply src/pages/iou/steps/MoneyRequestConfirmPage.js:155]: Yes 🙇 . Sorry for that.

File: src/pages/iou/steps/MoneyRequestConfirmPage.js
Comment (jeet-dhandha) [reply src/pages/iou/steps/MoneyRequestConfirmPage.js:158]: Slip of hand 🥲

**Comment by jeet-dhandha** [comment]
Screenshot 2023-09-02 at 6 00 50 AM
@hayata-suenaga  `CreateDistanceRequest` api returns this in response

**Comment by jeet-dhandha** [comment]
I have updated videos and code please check and get back.

**Comment by neil-marcellini** [comment]
@jeet-dhandha I will fix the backend to return reportID as a string, but it won't be deployed for a while. So please leave the handling to turn it into a string and a comment explaining that we will fix it on the backend. I will create a follow up issue.

**Comment by jeet-dhandha** [comment]
OKay @neil-marcellini no issues 👍

**Comment by jeet-dhandha** [comment]
Added comment

**Comment by jeet-dhandha** [comment]
Screenshot 2023-09-02 at 6 28 54 AM

It looks like a default rate.

**Comment by jeet-dhandha** [comment]
Screenshot 2023-09-02 at 6 30 50 AM

You can update that from workspace settings > reimbursement  > track distance

**Comment by neil-marcellini** [comment]
Distance requests use the workspace default rate. When you change the workspace currency under workspace settings it changes the currency in the rate. We want to use the currency in the rate.

If the rate is 0.625 USD and you change the currency to PKR we don't convert the rate, it's correctly 0.625 PKR.

**Comment by jeet-dhandha** [comment]
@allroundexperts Please ping if anything needed.

**Review by hayata-suenaga — APPROVED** [review]

**Comment by hayata-suenaga** [comment]
@allroundexperts please proceed with the rest of the checklist 🙇

**Review by allroundexperts — APPROVED** [review]
Looks good!

**Review by hayata-suenaga — APPROVED** [review]

File: src/components/MoneyRequestConfirmationList.js
Comment (hayata-suenaga) [inline src/components/MoneyRequestConfirmationList.js:296]: I don't like this pattern though.

We should get the distance request merchant string and set the string in IOU at the same place.

We can put the following code inside `useEffect` with the dependency array.

```ts
const distanceMerchant = DistanceRequestUtils.getDistanceMerchant(distance, unit, rate, currency, translate)
IOU.setMoneyRequestMerchant(distanceMerchant);
```

And for the `title` of `MenuItemWithTopDescription`, we can pass `iou.merchant` instead.

I gonna approve this PR right now, but please make a follow up PR @Santhosh-Sellavel for the cleanup 🙇 

We consider the issue to be complete after the follow-up PR is merged 👍

Comment (jeet-dhandha) [reply src/components/MoneyRequestConfirmationList.js:296]: @hayata-suenaga Okay.

File: src/libs/TransactionUtils.js
Comment (neil-marcellini) [reply src/libs/TransactionUtils.js:270]: You can test it out and see. Hopefully it has been fixed.

**Pull request merged**
